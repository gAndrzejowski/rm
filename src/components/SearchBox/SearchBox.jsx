// @flow
import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { searchMovies } from '../../actions/async';
import { setSearchTxt, setSearchCriterion } from '../../actions/creators';
import styles from './SearchBox.scss';
import type { Store } from '../../flowTypes';

type Props = {
  searchBy: string,
  query: string,
  onQueryChange: Event => void,
  onCriterionChosen: string => void,
  search: (string, string) => void
}

export const SearchBox = ({
  searchBy, query, search, onQueryChange, onCriterionChosen,
}: Props) => (
  <Fragment>
    <h2>Find your movie</h2>
    <input
      className={styles.searchInput}
      type="text"
      placeholder={searchBy === 'title' ? 'Kill Bill' : 'Action'}
      onChange={onQueryChange}
      value={query}
    />
    <div className={styles.controls}>
      <div className={styles.controls}>
                       Search by
        <button
          onClick={() => onCriterionChosen('title')}
          className={searchBy === 'title' ? styles.btnActive : styles.btnInactive}
        >title
        </button>
        <button
          onClick={() => onCriterionChosen('genres')}
          className={searchBy === 'genres' ? styles.btnActive : styles.btnInactive}
        >genre
        </button>
      </div>
      <Link href={`/search/${searchBy}/${query}`} to={`/search/${searchBy}/${query}`}>
        <button className={styles.search} onClick={() => search(query, searchBy)}>Search</button>
      </Link>
    </div>
  </Fragment>
);

SearchBox.defaultProps = {
  query: '',
};

export const mapStateToProps = (state :Store):{
  query: string,
  searchBy: string
} => ({
  query: state.heading.search,
  searchBy: state.heading.by,
});

export const mapDispatchToProps = (dispatch :any) => ({
  onQueryChange: (event: Object) => dispatch(setSearchTxt(event.target.value)),
  onCriterionChosen: (crit :string) => dispatch(setSearchCriterion(crit)),
  search: async (query :string, searchBy :string) => dispatch(await searchMovies(query, searchBy)),
});

export default connect(mapStateToProps, mapDispatchToProps)(SearchBox);
