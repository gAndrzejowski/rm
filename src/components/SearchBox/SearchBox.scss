@mixin button($bg) {
  color:white;
  text-transform: uppercase;
  text-align: center;
  display:block;
  background-color: $bg;
  margin-left:10px;
  border:none;
  padding:.2em 1ch;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  outline:none;
}
.btnActive,.search {
  @include button(firebrick);
}
.btnInactive {
  @include button(grey);
}
.controls {
  display:flex;
  justify-content: space-between;
  align-items:center;
}
.searchInput {
  width:100%;
  color:white;
  background:#000;
  padding:1em;
  border:none;
  border-bottom: 3px solid firebrick;
  margin-bottom:1em;
}