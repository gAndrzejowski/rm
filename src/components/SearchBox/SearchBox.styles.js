// @flow
const button = (bg :string) => ({
  cursor: 'pointer',
  color: 'white',
  textTransform: 'uppercase',
  textAlign: 'center',
  display: 'block',
  backgroundColor: bg,
  marginLeft: 10,
  border: 'none',
  padding: ['.2em', '1ch'],
  userSelect: 'none',
  outline: 'none',
});
const styles = {
  'btnActive, search': button('firebrick'),
  btnInactive: button('grey'),
  controls: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  searchInput: {
    width: '100%',
    color: 'white',
    background: '#000',
    padding: '1em',
    border: 'none',
    borderBottom: [3, 'solid', 'firebrick'],
    marginBottom: '1em',
  },
};

export default styles;
