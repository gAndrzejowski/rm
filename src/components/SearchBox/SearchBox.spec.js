import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import SearchBox from "./SearchBox";
import styles from './SearchBox.scss';

describe('SearchBox component', () => {
    it('renders correctly', () => {
        const tree = renderer.create(<SearchBox search={jest.fn()}/>).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('initializes state as empty query and search by title', () => {
        const wrapper = shallow(<SearchBox search={jest.fn()}/>);
        expect(wrapper.state()).toEqual({
            query:'',
            searchBy:'title'
        });
    })
    it('changes the state based on change events in UI components and updates their value', () => {
        const wrapper = shallow(<div><SearchBox search={jest.fn()}/></div>);
        const searchBox = wrapper.find('SearchBox').first().dive();

        searchBox.find(`.${styles.searchInput}`).simulate('change', {target:{value:'Machete'}});
        expect(searchBox.state('query')).toBe('Machete');
        expect(searchBox.find(`.${styles.searchInput}`).prop('value')).toBe('Machete');

        searchBox.find('button[onClick]').at(1).simulate('click');
        expect(searchBox.state('searchBy')).toBe('genre');
        expect(searchBox.find('button[onClick]').at(1).prop('className')).toBe(styles.btnActive);
        expect(searchBox.find('button[onClick]').at(0).prop('className')).toBe(styles.btnInactive);

        searchBox.find('button[onClick]').at(0).simulate('click');
        expect(searchBox.state('searchBy')).toBe('title');
        expect(searchBox.find('button[onClick]').at(0).prop('className')).toBe(styles.btnActive);
        expect(searchBox.find('button[onClick]').at(1).prop('className')).toBe(styles.btnInactive);
    })
    it('calls the search prop after clicking the search button, and passes the state values', () => {
        const wrapper = shallow(<div><SearchBox search={jest.fn()}/></div>);
        const searchBox = wrapper.find('SearchBox').first().dive();
        const mock =wrapper.find('SearchBox').prop('search').mock;
        const stateMock = {
            query:'action',
            searchBy:'genre'
        };

        searchBox.setState(stateMock);
        searchBox.find(`.${styles.search}`).simulate('click')

        expect(mock.calls[0][0]).toBe(stateMock.query);
        expect(mock.calls[0][1]).toBe(stateMock.searchBy);
    })
});