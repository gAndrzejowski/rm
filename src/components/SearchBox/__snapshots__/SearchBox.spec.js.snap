// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox component renders correctly 1`] = `
Array [
  <h2>
    Find your movie
  </h2>,
  <input
    className="searchInput"
    onChange={[Function]}
    placeholder="Kill Bill"
    type="text"
    value=""
  />,
  <div
    className="controls"
  >
    <div
      className="controls"
    >
      Search by
      <button
        className="btnActive"
        onClick={[Function]}
      >
        title
      </button>
      <button
        className="btnInactive"
        onClick={[Function]}
      >
        genre
      </button>
    </div>
    <button
      className="search"
      onClick={[Function]}
    >
      Search
    </button>
  </div>,
]
`;
